### DISCUSSION TITLE
TasmoHAB - OpenHAB Things and Items Generator via GUI

### DISCUSSION THREAD

**Discussion by Gifford47** [arendst/Tasmota#12499] [Show and tell]
Hi!
I would like to introduce my program https://github.com/Gifford47/tasmohab, which I use to integrate tasmota devices in openhab smart home systems.
A big thanks goes first to jimtng with his repo [jimtng/ohgen: OpenHAB Things and Items Generator](https://github.com/jimtng/ohgen). This made my work possible!
The program still has potential, but I hope it will be helpful to some.
I am still filling the readme, but I hope for active participation and improvements!

**Comment by ascillato** [comment]
Hi, your link is broken. No such repository (TasmoHab) in https://github.com/Gifford47

**Reply by meingraham** [reply]
Same issue for me.

**Reply by barbudor** [reply]
It's just the link that as a `"` at the end : https://github.com/Gifford47/tasmohab
@Gifford47 please edit your original post, thanks

**Reply by ascillato** [reply]
Now that repository is online and public. Some minutes ago there was only 7 repositories at https://github.com/Gifford47
Now there are 8

**Comment by ascillato** [comment]
@Gifford47

Please, can you add some information about this in Tasmota docs? Thanks.

**Reply by Gifford47** [reply]
> @Gifford47
> 
> Please, can you add some information about this in Tasmota docs? Thanks.

Sure! I'll try to add it in a few days. Has anyone already tried the program?

**Reply by barbudor** [reply]
I think @Jason2866 is an OpenHAB user and may want to have a look

**Comment by Jason2866** [comment]
Looks very interesting! Seems it solves the beginner hurdles to get Tasmota devices (with sensors) easily integrated in OpenHab. Will have a look the next days.

**Comment by sartyx** [comment]
I love the idea of the program, but why no binaries? At least pretty please sort out the documentation how to compile it.  Why is the convert to .exe listed under "Debugging"? I don´t want to complain too much here, the program probably is worth a try but noone is going to use it if the usage of it is a the same difficulty level as the direct editing in openHAB ;-) I just tried to compile it and just get a long list of errors :-/

**Reply by Jason2866** [reply]
Why not use with Python?

**Reply by sartyx** [reply]
Yes I tried that but after installing a bunch of modules it currently fails with PyQt6... I am on it but I am just too old for this I guess ;-)

**Comment by Jason2866** [comment]
Probably this `PyQt5>=5.15.6` in requirements.txt is not a good choice. A lot has changed in PyQt6. So latest PyQt5 is the one

**Reply by sartyx** [reply]
sigh yes, maybe but on both: error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/

16.11.14 is installed
Might not a problem for people using this stuff each day, I don´t, how many Tasmota/openHAB users do?
Of course this isn´t a tasmoHAB problem, it is more a pip problem I assume. Maybe I can install it somehow else.

EDIT: okay got it, leading to the next problem :

 ```
Traceback (most recent call last):
  File "tasmohab.py", line 24, in <module>
    from serial import Serial
ImportError: cannot import name 'Serial' from 'serial' (unknown location)

```

EDIT2: installing pyserial instead of serial fixes it ... finally

**Reply by Gifford47** [reply]
Why you didn't use the windows binary?

**Reply by sartyx** [reply]
Uh. Where? Anyway got it working now it just isn´t helping me much. Thought it would help me generate some switches for commands. But it might come handy with the next tasmota device.

**Reply by Gifford47** [reply]
> Uh. Where? Anyway got it working now it just isn´t helping me much. Thought it would help me generate some switches for commands. But it might come handy with the next tasmota device.

The binary is in the zip file from the last release.
If you use the table after getting tasmota device data, you can set the membership, tags, labels and additional openhab data. Then you create all appropriate things and items with just one click.
In my case it was very useful. It would be great if all users think so. So please let me know how I can improve the usage!

**Comment by sartyx** [comment]
Alright my fault, I didn´t notice the "Releases" I simply got the tasmohab-master. I have some difficulties to connect to the tasmota device. 
I assume the timeout is a bit early. Tasmohab will exit then in some cases. I guess that is because the wifi connection isn´t the best.
I had luck yesterday 1 in about 20 attempts though.

```
Traceback (most recent call last):
  File "tasmohab.py", line 309, in get_data_on_http
    resp_code = requests.get(self.http_url, timeout=.5).status_code
  File "C:\Users\tomqu\Anaconda3\lib\site-packages\requests\api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
requests.exceptions.ConnectTimeout: HTTPConnectionPool(host='192.168.178.102', port=80): Max retries exceeded with url: /cm?cmnd= (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x0000028614E30248>, 'Connection to 192.168.178.102 timed out. (connect timeout=0.5)'))
```

```
Traceback (most recent call last):
  File "tasmohab.py", line 239, in datathread_dev_data
    self.thing_id = json_dev_status[tas_cmds.status['network']]['StatusNET']['Hostname']           # set the global thingid
KeyError: 'status 5'
```

**Comment by sartyx** [comment]
Some suggestions for improvement:

- Most important fix the timeout thing, I am not sure what causes it but I am willing to find it out, if you let me know what I should change in code. The program shouldn´t crash of course.

- Edit Template calls a program which doesn´t exist, at least here. Not sure what this is but it looks like nothing is associated with .tpl and because of that it doesn´t open anything.

- At "Ip Address" in the GUI replace this with "Ip Address of your TASMOTA device". That prevents idiots like me trying the openHAB ip there if the TASMOTA ip fails to connect ;-)

- If the connections fails it asks to check user/password. That is a wrong error message in the case that the TASMOTA device doesn´t have any l/p set.

I can´t really test any further now, because I don´t get a connection again. However I noticed yesterday that trying to add the items to OH3 by "Adding Items from Textual Definition" end up in an error:

```
Error: Syntax error at line 1 col 1:

  [
  ^
Unexpected lbracket token: "[". Instead, I was expecting to see one of the following:
```

I never really used this before so I didn´t get any deeper into the problem by now. I assume you are using the files directly in the file structure of OpenHAB?
